cranelift/unwinder/src/stackwalk.rs

```diff
@@ -0,0 +1,230 @@
+
+    /// Debug assertion that the frame pointer is aligned.
+    fn assert_fp_is_aligned(&self, fp: usize);
+}
+
+/// A host-backed implementation of unwinding, using the native platform ABI
+/// that Cranelift has.
+#[cfg(has_host_compiler_backend)]
+pub struct UnwindHost;
+
+#[cfg(has_host_compiler_backend)]
+unsafe impl Unwind for UnwindHost {
+    fn next_older_fp_from_fp_offset(&self) -> usize {
+        arch::NEXT_OLDER_FP_FROM_FP_OFFSET
+    }
+    fn next_older_sp_from_fp_offset(&self) -> usize {
+        arch::NEXT_OLDER_SP_FROM_FP_OFFSET
+    }
+    unsafe fn get_next_older_pc_from_fp(&self, fp: usize) -> usize {
+        arch::get_next_older_pc_from_fp(fp)
+    }
+    fn assert_fp_is_aligned(&self, fp: usize) {
+        arch::assert_fp_is_aligned(fp)
+    }
+}
+
+/// An implementation specifically designed for unwinding Pulley's runtime stack
+/// (which might not match the native host).
+pub struct UnwindPulley;
+
+unsafe impl Unwind for UnwindPulley {
+    fn next_older_fp_from_fp_offset(&self) -> usize {
+        0
+    }
+    fn next_older_sp_from_fp_offset(&self) -> usize {
+        if cfg!(target_pointer_width = "32") {
+            8
+        } else {
+            16
+        }
+    }
+    unsafe fn get_next_older_pc_from_fp(&self, fp: usize) -> usize {
+        // The calling convention always pushes the return pointer (aka the PC
+        // of the next older frame) just before this frame.
+        *(fp as *mut usize).offset(1)
+    }
+    fn assert_fp_is_aligned(&self, fp: usize) {
+        let expected = if cfg!(target_pointer_width = "32") {
+            8
+        } else {
+            16
+        };
+        assert_eq!(fp % expected, 0, "stack should always be aligned");
+    }
+}
+
+/// A stack frame within a Wasm stack trace.
+#[derive(Debug)]
+pub struct Frame {
+    /// The program counter in this frame. Because every frame in the
+    /// stack-walk is paused at a call (as we are in host code called
+    /// by Cranelift code below these frames), the PC is at the return
+    /// address, i.e., points to the instruction after the call
+    /// instruction.
+    pc: usize,
+    /// The frame pointer value corresponding to this frame.
+    fp: usize,
+}
+
+impl Frame {
+    /// Get this frame's program counter.
+    pub fn pc(&self) -> usize {
+        self.pc
+    }
+
+    /// Get this frame's frame pointer.
+    pub fn fp(&self) -> usize {
+        self.fp
+    }
+}
+
+/// Walk through a contiguous sequence of Cranelift frames starting
+/// with the frame at the given PC and FP and ending at
+/// `trampoline_fp`. This FP should correspond to that of a trampoline
+/// that was used to enter the Cranelift code.
+///
+/// We require that the initial PC, FP, and `trampoline_fp` values are
+/// non-null (non-zero).
+pub unsafe fn visit_frames<R>(
+    unwind: &dyn Unwind,
+    mut pc: usize,
+    mut fp: usize,
+    trampoline_fp: usize,
+    mut f: impl FnMut(Frame) -> ControlFlow<R>,
+) -> ControlFlow<R> {
```

Comment:
If this function is already monomorphized over `R`, we might as well use an `&impl Unwind` instead of a `&dyn Unwind`

Reply:
Interestingly, this ends up propagating all the way up to e.g. `store.unwinder()` -- the `dyn` is infectious and we'd have a monomorphize a bunch more to make this work (or build a proxy-to-dyn impl to monomorphize on). Since this seems like just an opportunistic thing, I'll leave it as `dyn Unwind` for now...

Reply:
Yeah changing this away from `dyn Unwind`, while possible, I think is best to defer to later.